### ISSUE TITLE
Improve read-only behaviour for Markdown cells

### ISSUE THREAD

**Issue by agoose77** [jupyterlab/jupyterlab#9976]
### Problem
TL;DR
> The default view for read-only Markdown cells is a read-only CodeMirror editor. This is not particularly useful, and in some cases it might be useful to prevent the user from seeing the Markdown source.

A number of content authors are using Jupyter Notebooks as guided workbooks. These notebooks are often pre-populated with Markdown cells which guide the user through a particular set of tasks. When used in this manner, users are not expected to edit the Markdown cells, which should be read-only. 

The notebook specification defines metadata fields that render cells as non-deleteable and read-only. However, JupyterLab implements read-only views on cells as a read-only CodeMirror editor. In the context of the given use case, this means that users can accidentally double click on Markdown cells, which hides the rendered markup. This is not always a desired behaviour, and it would be nice if there were a mechanism by which authors could opt-in to disable this.

### Proposed Solution
@jasongrout proposed adding a setting to prevent showing a read-only cell content editor, and enabling this behaviour by default.

@bollwyvl suggested that one could add a "locked" icon with an `aria-label` to the active cell. This might be an orthogonal change to the setting suggested by @jasongrout.

Mockup:
![image](https://user-images.githubusercontent.com/1248413/111999482-f02cb280-8b14-11eb-9ee9-e4dbc2c6a593.png)

**Comment by jasongrout** [comment]
> @jasongrout proposed adding a setting to prevent showing a read-only cell content editor, and enabling this behaviour by default.

I'd even in be in favor of not having a setting, and just saying that read-only markdown cells do not ever change to show the source. If you want to see the source, first change the cell to editable. My (untested) guess is that in the vast majority of the cases, people accidentally double-click on markdown cells and do not want to see the source.

As a check: what does classic Notebook do?

**Comment by agoose77** [comment]
The behaviour in classic notebook matches that of JupyterLab

**Comment by goanpeca** [comment]
>The behaviour in classic notebook matches that of JupyterLab. Personally a 👍 on your suggestion to never show an editor for read-only Markdown cells.

👍🏼 on this!

**Comment by bollwyvl** [comment]
So, down in the guts:
- A: `packages/cells/src/widget.ts`
  - there's `rendered` and `readOnly`, both getter/setters
- B: `packages/notebook/src/widget.ts`
  - there's a `mode` which sets `rendered` iff it detects a `MarkdownCell`

does it make sense to: 

- A: intercept an update of `rendered = false` and replace it with `true` iff `readOnly`? 
- B: intercept (and disregard) `mode = 'edit'`

I imagine wanting the _same_ behavior and UI (whatever it ends up being) in other places like chat, pull-requests, etc.

I feel like making it _impossible_ to view the source (without gear -> accordion -> json editing) is very un-user-centric, and wouldn't work anyway (e.g. an instructor couldn't hide "answers" to quizzes, etc). So that leaves still needing a _View Source_:
- the lock (or maybe more accurate, strike-through-pencil) on the _active_ cell (probably in the prompt area?) 
  - if  _used_ to the current behavior, this would provide a visual cue
- hidden in the (already long) context menu (e.g. _View Source_)
  - comes with its own set of discoverability problems

**Comment by krassowski** [comment]
Things moving pretty fast here! I'm just catching up on the discussion but:
- let's add a command to lock/unlock all markdown editors in a particular notebook. This might be just a command that is hidden in the command palette (or could also be exposed as a button in the toolbar)
- I like the reasoning behind preventing the cells in read-only mode from entering the source editor (being un-rendered)
- a visual indicator might be useful if the command mentioned above is added (currently changing the cell metadata is slightly cumbersome process and unlikely to happen by accident, but once a command that switches the `editable` state is in, it is good to have at least some kind of visual indicator for users who do so accidentally)

**Comment by bollwyvl** [comment]
Here's the nickle fix (well, maybe 7.25 cents, to handle the invariant when the metadata changes externally):

```diff
diff --git a/packages/cells/src/widget.ts b/packages/cells/src/widget.ts
index c0c4b533c..af9712676 100644
--- a/packages/cells/src/widget.ts
+++ b/packages/cells/src/widget.ts
@@ -1418,6 +1418,9 @@ export class MarkdownCell extends AttachmentsCell<IMarkdownCellModel> {
     return this._rendered;
   }
   set rendered(value: boolean) {
+    if (this.readOnly) {
+      value = true;
+    }
     if (value === this._rendered) {
       return;
     }
@@ -1431,6 +1434,16 @@ export class MarkdownCell extends AttachmentsCell<IMarkdownCellModel> {
     }
   }
 
+  /**
+   * Load view editable state from model.
+   */
+  loadEditableState() {
+    super.loadEditableState();
+    if (this.readOnly) {
+      this.rendered = true;
+    }
+  }
+
   /**
    * Render an input instead of the text editor.
    */

```

The prompt area is... kind of a mess, on account of the default behavior manipulating `textContent`, so would be hard to change in a backwards-compatible way.

Also note, cell drag/drop is still enabled. If one is indeed expecting to double-click, it's _quite_ easy to accidentally start moving the cell. Should uneditable cells be draggable?

**Comment by bollwyvl** [comment]
Here's a potential treatment:

| remove the brand color from the collapser, just leave an empty box | ![Screenshot from 2021-03-23 00-44-32](https://user-images.githubusercontent.com/45380/112094074-f8b2d680-8b70-11eb-99d5-8b9bf765e2ea.png)
|-|-|
| or maybe some [hatch marks] that re-use theme variables? | ![Screenshot from 2021-03-23 00-56-33](https://user-images.githubusercontent.com/45380/112095321-a70b4b80-8b72-11eb-88fe-cd0e1c4e43ee.png)
| add a less-hidden metadata tool | ![Screenshot from 2021-03-23 00-45-30](https://user-images.githubusercontent.com/45380/112094170-2861de80-8b71-11eb-8bb8-e026f6125a81.png)

[hatch marks]: https://css-tricks.com/stripes-css/

**Comment by agoose77** [comment]
+1 for exposing the cell metadata in the toolbar. My gut instinct is to create a separate PR for those changes, as they're orthogonal?

RE dragging read-only cells, I think that might exceed the scope of the editable metadata. I can envisage a case where someone would want to make a cell read only, but is happy for it to remain dragged. For example, I'd probably use the editable flag more frequently in my literate-style notebooks, but I still like to be able to reorder things.

**Comment by bollwyvl** [comment]
> cell metadata in the toolbar

Oy, I'd rather core only add stuff to the toolbar very carefully... or it's going to become an MS Office ribbon. I guess a _single_ metadata button that popped the sidebar would be decent, but per-metadata-entry would rapidly get to be a lot.

> My gut instinct is to create a separate PR for those changes, as they're orthogonal?

I think any time a feature gets _taken away_ that used to work (even if, in this case, we're saying "double click to view source" _didn't ever work_) without a cue, it's an almost guaranteed set of future issues from confused users.

>  some refactoring to help encapsulate some of the functionality a little more carefully

My kingdom for xstate. _That_ would be a separate issue, for sure.

**Comment by bollwyvl** [comment]
(again, that's why the above are just inline patches, and not a PR, so we can _have_ these discussions without burning all the coal in kentucky)

**Comment by bollwyvl** [comment]
Semi-related, I'd always imagined an (optional) annotation sidebar where bespoke metadata, or perhaps, more humanely, "annotations" would go.

![Screenshot from 2021-03-23 10-41-55](https://user-images.githubusercontent.com/45380/112164623-6b996d00-8bc4-11eb-8709-950c5696c9a2.png)

This would be wonderful space for all kinds of things like commenting, viewing alternate mime types, tags, translations, etc. and would be more in line with direct manipulation.

**Comment by agoose77** [comment]
> > cell metadata in the toolbar
> 
> Oy, I'd rather core only add stuff to the toolbar very carefully... or it's going to become an MS Office ribbon. I guess a _single_ metadata button that popped the sidebar would be decent, but per-metadata-entry would rapidly get to be a lot.

I made a typo here, I meant the existing cell property inspector.

**Comment by bollwyvl** [comment]
Re: one of the original points: 

> adding a setting to prevent showing a read-only cell content editor, and enabling this behaviour by default.

this is probably the right play, but will increase the complexity of the setup a hair. Any thoughts on the naming things part?

maybe in `packages/notebook-extension/schema/tracker.json`:
```yaml
    "showEditorForReadOnlyMarkdown": {
      "title": "Show editor for read-only Markdown cells",
      "description": "Should an editor be shown for read-only markdown",
      "type": "boolean",
      "default": false
    }
```

**Comment by agoose77** [comment]
I have put my thoughts to the question and I cannot suggest a better name than `showEditorForReadOnlyMarkdown`, so that get's a :+1: from me.

**Comment by jasongrout** [comment]
`showEditorForReadOnlyMarkdown` sounds good to me. It's a little confusing, as it sounds like it won't be rendered, just shown in an editor, but I think the name is long enough and probably isn't that confusing in practice.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/create-an-extension-to-make-the-cell-un-editable-by-just-clicking-a-button/9282/2

**Comment by krassowski** [comment]
This was also recently mentioned in https://stackoverflow.com/questions/68869710/preventing-markdown-cells-from-un-rendering-jupyter. CC @krassowska who originally prompted me to start the preceding gitter discussion ([here](https://gitter.im/jupyterlab/jupyterlab?at=60566f5083533831b4f34637) and [here](https://gitter.im/jupyterlab/jupyterlab?at=605675d888edaa1eb8cd6acc)).

I think that we have three independent tasks here:
- [ ] add `showEditorForReadOnlyMarkdown` (default `false`) **and** simultaneously explore adding a visual cue for this setting being set to `true`. I like the lock icon idea as it quite prominent.
   - I would prefer only showing the lock icon on hover, in top-right corner,
   - the lock icon should not re-flow the rendered text (it should be above the rendered cell),
   - (as mentioned earlier it should actually be strike-through-pencil icon)
   - the lock icon could either have:
      - an on-click "unlock" action with title "This cell is read-only. Click to make it editable.":
         - \+ easy to use
         - \- if the lock would just disappear there would be no reverse action
      - an on-click context-menu/popover with options "Make this cell editable", and "Show editors for all read-only markdown cells":
         - \+ as it avoids accidental clicks
         - \+ can add more options in the future
         - \- extra complexity
- [ ] add hatches to signal that a cell is locked, independent of the `showEditorForReadOnlyMarkdown` state
- [ ] promote "readonly" to its own field in the existing cell metadata inspector

I see this as three independent PRs. Does this plan sound good?

**Comment by krassowski** [comment]
Also, any preference towards the single on-click action vs having a context-menu like pop-over with a few actions to choose from? CC @isabela-pf?

**Comment by agoose77** [comment]
* :+1: on promoting `readonly` to a cell-toolbar widget. In my view, anything *standardised* in the notebook schema should be rendered.
* I think the idea of a lock icon is good, but I would suggest a different use - as a `:focused { visibility: visible; }`  for read-only cells (independent of the `showEditorForReadOnlyMarkdown` proposal). This way, we don't clutter the UI, as (in my opinion) read-only only matters when you focus a cell and try to edit it.  
Here's a lazy mockup to make the point:
![Peek 2021-08-22 14-40](https://user-images.githubusercontent.com/1248413/130357328-79b65826-1cf9-4cf2-9f41-adc0439940ad.gif)
* Now, `showEditorForReadOnlyMarkdown` just disables the ability to enter edit mode for read-only (markdown) cells. Maybe the inverted `onlyRenderReadOnlyMarkdown` would be better?

**Comment by krassowski** [comment]
Folks do care about the overall look and feel of the notebook so I am not a fan of having the icon take up vertical space. Taking up horizontal space would be better, but still awkward because cells would end up having different width. My preference remains to have it on hover. I like the idea of showing a cue on focus. We could manipulate the opacity:
- focus: 25% (text underneath is still visible)
- hover on cell: 50% (it becomes more prominent, but does not obstruct the view)
- hover on icon: 100%

No strong opinion on `onlyRenderReadOnlyMarkdown`, but it feels more difficult to mentally parse for me than `showEditorForReadOnlyMarkdown`.

**Comment by agoose77** [comment]
> Folks do care about the overall look and feel of the notebook so I am not a fan of having the icon take up vertical space. Taking up horizontal space would be better, but still awkward because cells would end up having different width. My preference remains to have it on hover. I like the idea of showing a cue on focus. We could manipulate the opacity:
> 
> * focus: 25% (text underneath is still visible)
> * hover on cell: 50% (it becomes more prominent, but does not obstruct the view)
> * hover on icon: 100%
> 
> No strong opinion on `onlyRenderReadOnlyMarkdown`, but it feels more difficult to mentally parse for me than `showEditorForReadOnlyMarkdown`.

Sure, the mockup is only to emphasise "visible-on-active". I'd probably suggest putting the icon in the top right corner of the cell itself (overlay). I like the idea of a phased visibility. That would work better with the icon overlaid over the cell, rather than requiring special padding.

**Comment by behrouz-bakhtiari** [comment]
Hi,
I really like the idea of the lock icon on hover. Just wanted to mention that right now the code cells are still clickable even when "editable": false is added. You realize that it is not editable when you try to erase or add something, not before. So a non-editable markdown cell could also potentially behave the same way, for consistency reasons. But of course adding more visual cues are always good.

**Comment by krassowski** [comment]
Thanks @balamuruganvg, you are raising a very good point - we should decide if we want the lock (/pencil) icon to apply to markdown cells only, or to the code cells as well (not necessarily in the first PR which would implement it, but in the future).

For the non-editable code cells, in addition to the lock icon, we could also hide the cursor by applying `CodeMirror-readonly` class and changing the styling of the focused editor as well (I am not sure if we do anything like that already).

**Comment by behrouz-bakhtiari** [comment]
> Thanks @balamuruganvg, you are raising a very good point - we should decide if we want the lock (/pencil) icon to apply to markdown cells only, or to the code cells as well (not necessarily in the first PR which would implement it, but in the future).
> 
> For the non-editable code cells, in addition to the lock icon, we could also hide the cursor by applying `CodeMirror-readonly` class and changing the styling of the focused editor as well (I am not sure if we do anything like that already).

That's right. In fact, I realized that I could make code cells read only just yesterday and it already felt "more like a hack" rather than a standard procedure, exactly because there are no visual cues indicating a code cell is now read-only. So I am all for keeping the code and markdown cell behaviour consistent for read-only cells.

**Comment by isabela-pf** [comment]
Thanks for your patience with my slow reply! 

> Also, any preference towards the single on-click action vs having a context-menu like pop-over with a few actions to choose from? CC @isabela-pf?

I haven't dealt much with read-only notebooks, so I'm happy to weigh in but super open to hearing the thoughts of people who deal with them more. From reading this issue, it sounds like the on-click approach is more like what we are currently using and what is causing accidental interactions. Adding the lock as feedback will be an improvement regardless, and I think the context-menu could prevent the kind of accidents you all have described above.

> RE the latter proposals, I don't know what the accessibility ramifications are. I'm sure that @isabela-pf will have some thoughts? :)

For better or for worse, that is often the case! 😆 I'm not totally certain which proposal you are asking for feedback on, but I think it might be the accessibility ramifications of having a non-persistent lock icon? 

My best guess is that visually this is fine since whether or not a cell is read-only only matters when you are interacting with it anyways. Generally I've heard its better to have feedback like this be persistent, but I also understand how redundant that information could easily be with a lot of cells in a notebook. For making sure that this can effectively be non-visual feedback, I think this non-persistent icon might still work since I don't think it needs to be read without the cell being focused. As long as focus, and not just a mouse/hover/whatever, is what triggers the lock icon and related feedback it seems reasonable.

Keep in mind I used the word "guess." This isn't a pattern I've heard specifically discussed before or tried to fumble through with a screen reader myself.

**Comment by krassowski** [comment]
I don't recollect making a draft PR for this one ;)

**Comment by Pugio** [comment]
It seems to be listed just above: https://github.com/jupyterlab/jupyterlab/pull/10956

**Comment by agoose77** [comment]
> It seems to be listed just above: #10956

Ah, that's @krassowska !

**Comment by krassowska** [comment]
Hi @Pugio,

It requires adding some visual indication that the cell is in read-only mode. I do not think there was anything else to do.

I'm not sure why it's failing on the "Linux JS Flaky Tests / JS (js-apputils)" and "UI Tests / Visual Regression Tests".

Thanks, @agoose77 for tagging me.

**Comment by Pugio** [comment]
Thank you everyone, and apologies to the krassowsk*s.
Reading the above discussion, it seems to make the most sense to have the visual indicator only when a cell is selected; too visually noisy and largely irrelevant otherwise.

I suggest the following: (shown on multiple cells here only for visual inspection)
![image](https://user-images.githubusercontent.com/286180/149720149-3629b9f8-5feb-481f-9f5a-a0703e041987.png)

This could be accomplished with just adding this extra css rule:

```css
.jp-mod-selected .lm-Widget.p-Widget.jp-RenderedHTMLCommon.jp-RenderedMarkdown.jp-MarkdownOutput::before {
    content: '\f023';
    font-family: 'Font Awesome 5 Free';
    color: var(--jp-layout-color3);
    -webkit-font-smoothing: antialiased;
    font-style: normal;
    font-variant: normal;
    text-rendering: auto;
    font-weight: 600;
    position: absolute;
    right: 0;
}
```

**Comment by agoose77** [comment]
I think this has been resolved by #10956

**Comment by krassowski** [comment]
We are still missing read-only indicator though.

**Comment by krassowski** [comment]
Continuing discussion in https://github.com/jupyterlab/jupyterlab/issues/19051 - please chime in with your thoughts.
